gremlin-dotnet/src/Gremlin.Net/Driver/ConnectionPool.cs

```diff
@@ -113,32 +112,44 @@ private async Task<Connection> CreateNewConnectionAsync()
 
         private Connection GetConnectionFromPool()
         {
-            while (true)
-            {
-                var connection = SelectLeastUsedConnection();
-                if (connection == null)
-                    throw new ServerUnavailableException();
-                if (connection.NrRequestsInFlight >= _maxInProcessPerConnection)
-                    throw new ConnectionPoolBusyException(_poolSize, _maxInProcessPerConnection);
-                if (connection.IsOpen) return connection;
-                DefinitelyDestroyConnection(connection);
-            }
+            var connections = _connections.Snapshot;
+            if (connections.Length == 0) throw new ServerUnavailableException();
+
+            if (GetAvailableConnection(connections, out var connection))
+                return connection;
+            throw new ConnectionPoolBusyException(_poolSize, _maxInProcessPerConnection);
         }
 
-        private Connection SelectLeastUsedConnection()
+        private bool GetAvailableConnection(Connection[] connections, out Connection connection)
         {
-            if (_connections.IsEmpty) return null;
-            var nrMinInFlightConnections = int.MaxValue;
-            Connection leastBusy = null;
-            foreach (var connection in _connections)
+            var index = Interlocked.Increment(ref _connectionIndex);
+            ProtectIndexFromOverflowing(index);
+
+            connection = null;
+            for (var i = 0; i < connections.Length; i++)
             {
-                var nrInFlight = connection.NrRequestsInFlight;
-                if (nrInFlight >= nrMinInFlightConnections) continue;
-                nrMinInFlightConnections = nrInFlight;
-                leastBusy = connection;
+                connection = connections[(index + i) % connections.Length];
+                if (connection.NrRequestsInFlight >= _maxInProcessPerConnection) continue;
```

Comment:
Note that `NrRequestsInFlight` uses `ConcurrentDictionary.Count` which is super expensive.

We can maintain a separate counter or maybe limit the comparison to 2 connections max for now (and fix later): `for (...; i < Math.Min(connections.Length, 2); ...)`

Reply:
Do you think that that is really a problem? I expect the check to succeed for the first connection already in like 99,9% of the cases because thanks to the round-robin we should have an equal distribution of requests over connections and if the connections actually start reaching this limit, then it won't take long until a `ConnectionPoolBusyException` is thrown at which case users have to increase the limits any way which will make the situation again less likely to occur.
That's why I also don't see the benefit in limiting the number of checks here. It would just reduce the time until an exception is thrown minimally, but it probably won't result in performance benefits for 99,9% of the requests.

I can add a dedicated counter though to the `Connection` object if you want to. Just wanted to better understand your reasoning here.

Reply:
oh, nvm, I was thinking you were comparing connections with least in-flights and loop through all connections... sorry.